query-projector/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+            }
+        }
+    }
+
+    templates
+}
+
+pub trait Projector {
+    fn project<'stmt>(&self, rows: Rows<'stmt>) -> Result<QueryResults>;
+}
+
+struct ScalarProjector {
+    template: TypedIndex,
+}
+
+impl ScalarProjector {
+    fn with_template(template: TypedIndex) -> ScalarProjector {
+        ScalarProjector {
+            template: template,
+        }
+    }
+}
+
+impl Projector for ScalarProjector {
+    fn project<'stmt>(&self, mut rows: Rows<'stmt>) -> Result<QueryResults> {
+        if let Some(r) = rows.next() {
+            let row = r?;
+            let binding = self.template.lookup(&row)?;
+            Ok(QueryResults::Scalar(Some(binding)))
+        } else {
+            Ok(QueryResults::Scalar(None))
+        }
+    }
+}
+
+/// A tuple projector produces a single vector. It's the single-result version of rel.
+struct TupleProjector {
+    len: usize,
+    templates: Vec<TypedIndex>,
+}
+
+impl TupleProjector {
+    fn with_templates(len: usize, templates: Vec<TypedIndex>) -> TupleProjector {
+        TupleProjector {
+            len: len,
+            templates: templates,
+        }
+    }
+
+    // This is exactly the same as for rel.
+    fn collect_bindings<'a, 'stmt>(&self, row: Row<'a, 'stmt>) -> Result<Vec<TypedValue>> {
+        assert_eq!(row.column_count(), self.len as i32);
+        self.templates
+            .iter()
+            .map(|ti| ti.lookup(&row))
+            .collect::<Result<Vec<TypedValue>>>()
+    }
+}
+
+impl Projector for TupleProjector {
+    fn project<'stmt>(&self, mut rows: Rows<'stmt>) -> Result<QueryResults> {
+        if let Some(r) = rows.next() {
+            let row = r?;
+            let bindings = self.collect_bindings(row)?;
+            Ok(QueryResults::Tuple(Some(bindings)))
+        } else {
+            Ok(QueryResults::Tuple(None))
+        }
+    }
+}
+
+/// A rel projector produces a vector of vectors.
+/// Each inner vector is the same size, and sourced from the same columns.
+/// One inner vector is produced per `Row`.
+/// Each column in the inner vector is the result of taking one or two columns from
+/// the `Row`: one for the value and optionally one for the type tag.
+struct RelProjector {
+    len: usize,
+    templates: Vec<TypedIndex>,
+}
+
+impl RelProjector {
+    fn with_templates(len: usize, templates: Vec<TypedIndex>) -> RelProjector {
+        RelProjector {
+            len: len,
+            templates: templates,
+        }
+    }
+
+    fn collect_bindings<'a, 'stmt>(&self, row: Row<'a, 'stmt>) -> Result<Vec<TypedValue>> {
+        assert_eq!(row.column_count(), self.len as i32);
+        self.templates
+            .iter()
+            .map(|ti| ti.lookup(&row))
+            .collect::<Result<Vec<TypedValue>>>()
+    }
+}
+
+impl Projector for RelProjector {
+    fn project<'stmt>(&self, mut rows: Rows<'stmt>) -> Result<QueryResults> {
+        let mut out: Vec<Vec<TypedValue>> = vec![];
+        while let Some(r) = rows.next() {
```

Comment:
Two reasons:

- Some find specs/projectors — as you can see in `ScalarProjector` — will only ever consume one row. Rather than having two code paths, we simply always pass `Rows` to the projector, which can choose how much to consume.
- Some future projections will decouple the input rows from the output results, accumulating 0–N results for each SQL row; some of those behaviors will require context (such as accumulating into a `Set`). `query_map` wouldn't be appropriate for those.